amethyst_renderer/src/mesh.rs

```diff
@@ -1,44 +1,181 @@
+
+
+/// Raw buffer with its attributes
+#[derive(Clone, Debug)]
+pub struct VertexBuffer {
+    attrs: Attributes<'static>,
+    raw: RawBuffer,
+}
+
+/// Vertex data that can be built into `VertexBuffer`
+#[doc(hidden)]
+pub trait VertexData {
+    const ATTRIBUTES: Attributes<'static>;
+
+    /// Get vertex count in buffer
+    fn len(&self) -> usize;
+
+    /// Build `VertexBuffer`
+    fn build(&self, factory: &mut Factory) -> Result<VertexBuffer>;
+}
+
+/// Construct new vertex data from raw data and vertex format
+pub fn vertex_data<D, V>(data: D) -> (D, PhantomData<V>)
+where
+    D: AsRef<[V]>,
+    V: VertexFormat,
+{
+    (data, PhantomData)
+}
+
+impl<D, V> VertexData for (D, PhantomData<V>)
+where
+    D: AsRef<[V]>,
+    V: VertexFormat,
+{
+    const ATTRIBUTES: Attributes<'static> = V::ATTRIBUTES;
+
+    fn len(&self) -> usize {
+        self.0.as_ref().len()
+    }
+
+    fn build(&self, factory: &mut Factory) -> Result<VertexBuffer> {
+        use gfx::Factory;
+        use gfx::memory::{cast_slice, Bind};
+        use gfx::buffer::Role;
+
+        let verts = self.0.as_ref();
+        let slice = cast_slice(verts);
+        let stride = slice.len() / verts.len();
+        let role = Role::Vertex;
+        let bind = Bind::empty();
+
+        let vbuf = factory.create_buffer_immutable_raw(slice, stride, role, bind)?;
+        Ok(VertexBuffer {
+            attrs: V::ATTRIBUTES,
+            raw: vbuf,
+        })
+    }
+}
+
+/// Set of vertex data
+#[doc(hidden)]
+pub trait VertexDataSet {
+    /// Iterator for `VertexBuffer`s built
+    type VertexBufferIter: Iterator<Item = VertexBuffer>;
+
+    /// Get smalles vertex count across buffers
+    fn len(&self) -> usize;
+
+    /// Build `VertexBuffer`s
+    fn build(&self, factory: &mut Factory) -> Result<Self::VertexBufferIter>;
+}
+
+impl<H> VertexDataSet for (H, ())
+where
+    H: VertexData,
+{
+    type VertexBufferIter = Once<VertexBuffer>;
+
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    fn build(&self, factory: &mut Factory) -> Result<Self::VertexBufferIter> {
+        let (ref head, _) = *self;
+        Ok(once(head.build(factory)?))
+    }
+}
+
+impl<H, T> VertexDataSet for (H, T)
+where
+    H: VertexData,
+    T: VertexDataSet,
+{
+    type VertexBufferIter = Chain<Once<VertexBuffer>, T::VertexBufferIter>;
+
+    fn len(&self) -> usize {
+        use std::cmp::min;
+        min(self.0.len(), self.1.len())
+    }
+
+    fn build(&self, factory: &mut Factory) -> Result<Self::VertexBufferIter> {
+        let (ref head, ref tail) = *self;
+        Ok(once(head.build(factory)?).chain(tail.build(factory)?))
+    }
+}
+
 
 /// Represents a polygonal mesh.
-#[derive(Clone, Debug, PartialEq)]
+#[derive(Clone, Debug)]
 pub struct Mesh {
-    attrs: &'static [(&'static str, AttributeFormat)],
-    prim: Primitive,
     slice: Slice,
     transform: Matrix4<f32>,
-    vbuf: RawBuffer,
+    vbufs: SmallVec<[VertexBuffer; 4]>,
 }
```

Comment:
4 is gonna be too low later on when we start looking at animation and morphing. For skeletal animation you need atleast bone ids and bone weights per vertex.

Reply:
4 here isn't a limit. You can push more. They just will be on heap.

Reply:
Still, we should be sure to handle the common cases. I think we're using too much `SmallVec` in our crate; I've seen a benchmark in the past which showed that you shouldn't use `SmallVec` in most cases. Cannot remember where though.

Reply:
SmallVec is optimized for a very specific use case.  It should only be used when, for the overwhelming majority of the time, your collection will be under that many elements.  For example SmallVec works well for the input system because It is exceedingly rare that more than 8 keys will be pressed simultaneously.  It also works well for the WindowMessage system because it is incredibly unlikely you'll receive more than one WindowMessage per frame.  I don't think this however is an appropriate use case for SmallVec, as it sounds like the need for more VertexBuffers is going to grow quite rapidly.

Reply:
Actually there is no difference what to use here `Vec` or `SmallVec`. Because each `push` is associated with creating of `gfx::RawBuffer` which is much slower operation.

Reply:
Rapidly growing demand of vertex buffers isn't a good thing.
3 buffers with `[f32; 3]`, 1 with `[f32; 2]` will be the most case for static objects.
Additional 4 pairs of `u16` and `f32` for skeletal animation.
And that's it for 99.9% of meshes.

Reply:
If so, then 4 is fine